runtime/cudaq/spin/spin_op.cpp

```diff
@@ -112,116 +208,63 @@ void spin_op::for_each_pauli(
+    for (std::size_t i = 0; i < newSize / 2; i++) {
+      tmp.insert(tmp.begin() + tmp.size() / 2, 0);
+      tmp.insert(tmp.begin() + tmp.size(), 0);
+    }
 
-spin_op::spin_op(pauli type, const std::size_t idx,
-                 std::complex<double> coeff) {
-  m_n_qubits = idx + 1;
-  [[maybe_unused]] int p = 0;
-  data.push_back(std::vector<bool>(2 * m_n_qubits));
-  if (type == pauli::X) {
-    data.back()[idx] = true;
-  } else if (type == pauli::Y) {
-    p++;
-    data.back()[idx] = true;
-    data.back()[idx + m_n_qubits] = true;
-  } else if (type == pauli::Z) {
-    data.back()[idx + m_n_qubits] = true;
+    terms.erase(iter++);
+    terms.emplace(tmp, coeff);
   }
-
-  coefficients.push_back(coeff);
 }
 
-spin_op::spin_op(const spin_op &o)
-    : data(o.data), coefficients(o.coefficients), m_n_qubits(o.m_n_qubits) {}
-
 spin_op &spin_op::operator+=(const spin_op &v) noexcept {
-  spin_op tmpv = v;
-  if (v.m_n_qubits > m_n_qubits) {
-    // If we are adding a op that has more qubits than we do
-    // then we need to resize, making sure to ensure the
-    // correct 1/0 positions.
-    expandToNQubits(v.m_n_qubits);
-  } else if (v.m_n_qubits < m_n_qubits) {
-    tmpv.expandToNQubits(m_n_qubits);
-  }
-
-  // Add the rows from v to this, if
-  // the row already exists, we should just add the coeffs
-  for (auto [i, row] : enumerate(tmpv.data)) {
-    auto it = std::find(data.begin(), data.end(), row);
-    if (it != data.end()) {
-      auto idx = std::distance(data.begin(), it);
-      coefficients[idx] += tmpv.coefficients[i];
-    } else {
-      data.push_back(row);
-      coefficients.push_back(tmpv.coefficients[i]);
-    }
-  }
-
-  // mark any rows for deletion if coeff = (0,0)
-  std::vector<int> marked;
-  for (std::size_t i = 0; i < data.size(); i++)
-    if (std::abs(coefficients[i]) < 1e-12)
-      marked.push_back(i);
+  auto otherNumQubits = v.num_qubits();
 
-  std::sort(marked.begin(), marked.end(), std::greater<int>());
-  for (auto m : marked) {
-    data.erase(data.begin() + m);
-    coefficients.erase(coefficients.begin() + m);
+  spin_op tmpv = v;
+  if (otherNumQubits > num_qubits())
+    expandToNQubits(otherNumQubits);
+  else if (otherNumQubits < num_qubits())
+    tmpv.expandToNQubits(num_qubits());
+
+  for (auto [term, coeff] : tmpv.terms) {
+    auto iter = terms.find(term);
+    if (iter != terms.end())
+      iter->second += coeff;
+    else
+      terms.emplace(term, coeff);
   }
 
   return *this;
 }
 
-std::vector<std::complex<double>> spin_op::get_coefficients() const {
-  return coefficients;
-}
-
 spin_op spin_op::operator[](const std::size_t term_idx) const {
-  std::vector<bool> term_data = data[term_idx];
-  auto term_coeff = coefficients[term_idx];
-  BinarySymplecticForm f{term_data};
-  std::vector<std::complex<double>> c{term_coeff};
-  return spin_op(f, c);
+  auto start = terms.begin();
+  std::advance(start, term_idx);
+  return spin_op(start->first, start->second);
```

Comment:
We should remove this overload if the underling data structure is an `unordered_map`

Reply:
I will update to provide `begin/end` iterator functions.